runtime/Java/src/org/antlr/v4/runtime/tree/Trees.java

```diff
@@ -189,6 +198,107 @@ else if ( !findTokens && t instanceof ParserRuleContext ) {
 		return getDescendants(t);
 	}
 
+	/** Return a new TerminalNode with a copy of all relevant fields
+	 *  from t and with null parent.
+	 *
+	 *  @since 4.6.1
+	 */
+	public static TerminalNodeImpl shallowCopy(TerminalNode t) {
+		return new TerminalNodeImpl(t.getSymbol());
+	}
+
+	/** Create a shallow copy of node t; the copy has same type as t. Copy
+	 *  all relevant fields to make a proper clone of t except
+	 *  the children list. The only caveat is that the error nodes from t
+	 *  are in fact copied into the clone.
+	 *
+	 *  t is not altered.
+	 *
+	 *  Because this uses reflection, it might be slower than using
+	 *  "new X()" directly.  It uses {@see ctorCache} to speed things up a lot.
+	 *
+	 *  @since 4.6.1
+	 */
+	public static ParserRuleContext shallowCopy(ParserRuleContext t) {
+		try {
+			Class<? extends ParserRuleContext> cl = t.getClass();
+			Class<?> sup = cl.getSuperclass();
+			Class<?> supSup = sup.getSuperclass();
+			if ( ctorCache==null ) { // in a multithreaded environment, some thread will win to set ctorCache
+				ctorCache =	new ConcurrentHashMap<>();
+			}
+			boolean
+				isAltLabelNode = sup!=ParserRuleContext.class && // usual case
+				supSup!=ParserRuleContext.class;                 // if we use contextSuperClass option we are 2 below ParserRuleContext
+			if ( isAltLabelNode ) {
+				// AltLabel: nodes public UnarySuffixContext(StuffContext ctx) { copyFrom(ctx); }
+				Constructor<? extends ParserRuleContext> ctor = ctorCache.get(cl);
+				if ( ctor==null ) {
+					ctor = cl.getConstructor(sup); // no need for explicit copyFrom
+				}
+				return ctor.newInstance(t);
+			}
+			else {
+				// RuleNodes: public StatContext(ParserRuleContext parent, int invokingState) {
+				Constructor<? extends ParserRuleContext> ctor = ctorCache.get(cl);
+				if ( ctor==null ) {
+					ctor = cl.getConstructor(ParserRuleContext.class, int.class);
+					ctorCache.put(cl, ctor);
+				}
+				ParserRuleContext copy = ctor.newInstance(t.getParent(), t.invokingState);
+				copy.copyFrom(t); // make sure we copy all necessary fields (and possibly error node children)
+				return copy;
+			}
+		}
+		catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	/** Make a shallow copy of a generic parse tree node t
+	 *
+	 *  @since 4.6.1
+	 */
+	public static ParseTree shallowCopy(ParseTree t) {
+		if ( t instanceof TerminalNode ) return shallowCopy((TerminalNode)t);
+		return shallowCopy((ParserRuleContext)t);
+	}
+
+	/** Make a complete copy of a parse tree rooted at t.
+	 *  The original tree, t, is not altered.
+	 *
+	 *  @since 4.6.1
+	 */
+	public static ParseTree deepCopy(ParseTree t) {
```

Comment:
:bulb: I would leave this out for now.

1. It's not needed to get rewriting working
2. It fails to assign deep-copied nodes back to fields created for labels, which could lead to us fielding bug reports for it